Move meta-programming guide here from elixir-lang.org

whatyouhide opened the pull request:
As discussed with @josevalim and team.

This PR only moves the guides *as is*, without changes to the content of the guides themselves. I only removed the `---` YAML comments from the top of those files. This is so we don't have to review the files themselves 👍 

This is part of #12747.

josevalim commented on lib/elixir/pages/meta-programming/quote-and-unquote.md:
```diff
@@ -0,0 +1,154 @@
+# Meta-programming
+
+This guide aims to introduce the meta-programming techniques available in Elixir. The ability to represent an Elixir program by its own data structures is at the heart of meta-programming. This chapter starts by exploring those structures and the associated `quote` and `unquote` constructs, so we can take a look at macros in the next chapter and finally build our own domain specific language.
+
+> The Elixir guides are also available in EPUB format:
+>
+>   * [Getting started guide](https://repo.hex.pm/guides/elixir/elixir-getting-started-guide.epub)
+>   * [Mix and OTP guide](https://repo.hex.pm/guides/elixir/mix-and-otp.epub)
+>   * [Meta-programming guide](https://repo.hex.pm/guides/elixir/meta-programming-in-elixir.epub)
```
```suggestion
```

josevalim commented on lib/elixir/pages/meta-programming/quote-and-unquote.md:
```diff
@@ -0,0 +1,154 @@
+
+In other words, it is important to make a distinction between a regular Elixir value (like a list, a map, a process, a reference, etc) and a quoted expression. Some values, such as integers, atoms, and strings, have a quoted expression equal to the value itself. Other values, like maps, need to be explicitly converted. Finally, values like functions and references cannot be converted to a quoted  ...
+
+You can read more about `quote` and `unquote` in the [`Kernel.SpecialForms` module](https://hexdocs.pm/elixir/Kernel.SpecialForms.html). Documentation for `Macro.escape/1` and other functions related to quoted expressions can be found in the [`Macro` module](https://hexdocs.pm/elixir/Macro.html).
```
We no longer need to write Markdown links. :) The benefit of making it part of the documentation is that auto-linking will just work, so we can write this:

```suggestion
You can read more about `quote/1` and `unquote/1` in the `Kernel.SpecialForms` module. Documentation for `Macro.escape/1` and other functions related to quoted expressions can be found in the `Macro` module.
```

Can you please do a double pass in the docs and make sure to remove all hexdocs links in favor of just using markdown code syntax?

whatyouhide replied:
Yes sure, as I said in the PR I wanted to do this later on so that we can review it in isolation but I can do that here too 👍

josevalim replied:
Ah, sorry. I have reviewed the contents themselves several times over the years, so I am ready to review the contents too. For the other main guide, I want to port it myself, because I want to rewrite some larger sections. :)

whatyouhide commented:
@josevalim I went through all three guides and cleaned them up to fit better inside Elixir. Also removed external links and stuff like that. Ready for another review.

josevalim commented on lib/elixir/scripts/elixir_docs.exs:
```diff
@@ -28,6 +31,7 @@ canonical = System.fetch_env!("CANONICAL")
   ],
   groups_for_extras: [
     "Anti-patterns": ~r"pages/anti-patterns/.*\.md$",
+    "Meta-programming": ~r"pages/meta-programming/.*\.md$",
```
I am wondering if this should come after references but anyway works for me. :)

whatyouhide replied:
I think we can figure out a good order after we move all the guides really 🙃

whatyouhide commented:
Part of 12747.
